README.md
# Pokemon327
Terrain Generator makes different maps with two paths and has randomly generated terrain and 2 regions of tall grass and 2 regions of clearings.

CHANGELOG.txt
1/29/22: Created path builder to help connect exits
1/31/22: Created function to add Pokemon Cart and Pokemon Mart. Also added function to add grass,clearing,tree and boulder regions
2/1/22: Added bug fixes to path to require minimum distances between paths to 2. Also a bug fix for Mart and Center placement to prevent it from being placed
over path. There is a possible bug for the mart where it may place the mart on the N->S path but that is a rare occurence.
2/6/22: Created function to change maps on N S E W coordinates. Refactored TerrainGeneration.h to allow coordinate inputs so the exits are randomly generated in Map2Map.c instead of TerrainGeneration.h
2/8/22: Added border functionality when generating new maps. Also made sure that it blocked of two sides if the map is generate in the corner of the map. Created neighborCheck function a way of trying to connect newly generated maps with pre-existing maps.
Added integer values to the singular map struct so that exit locations can be returned from pre-existing maps.
Added flight functionality but only works from hardcoded values instead of user input.


TerrainGeneration.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

typedef struct table_t{
char map[21][80];
}table_t;
void mapbuilder(table_t *m){
  int i,j;
  for(i = 0; i < 21; i++){
      for(j = 0; j < 80; j++){
          if(i == 0|| i == 20 || j == 0|| j == 79)
          {
              m->map[i][j] = '%';

          }else{
              m->map[i][j] = ' ';
          }
      }
  }
}

void pathbuilder(table_t *m,int N,int S,int E,int W,int ranX,int ranY){

  int x_diff = E - W;
  int y_diff= N - S;
  //printf("X difference: %d, Y difference: %d\n",x_diff,y_diff);
  //printf("N:%d, S:%d\n",N,S);
  //printf("E:%d, W:%d, ranY: %d\n",E,W,ranY);
  m ->map[0][N] = '#';
  m ->map[20][S] = '#';
  m ->map[E][0] = '#';
  m ->map[W][79] = '#';
  int k = 79;
  int i,j;
  for(i = 0; i < 21; i++){
      for(j = 0; j < 80; j++){
          if(i == E && j <= ranX && j > 0){
              m->map[E][j] = '#';
          }
          if(E > W &&  x_diff>= 0){
              m ->map[E-x_diff][ranX] = '#';
              x_diff--;
          }else if(E < W && x_diff <= 0){
              m ->map[E-x_diff][ranX] = '#';
              x_diff++;
          }
          if(k > ranX)
          {
              m ->map[W][k] = '#';
              k--;
          }
      }
  }
  int f = 20;
  for(i = 0; i < 21; i++){
      for(j = 0; j < 80; j++){
          if(j == N && i <= ranY && i > 0){
               m ->map[i][N] = '#';
          }

          if(N > S && y_diff > 0){
               m ->map[ranY][N-y_diff] = '#';
                 y_diff--;

             }else if(N < S && y_diff < 0){
               m ->map[ranY][N-y_diff] = '#';
                 y_diff++;
             }

          if(f > ranY){
            m ->map[f][S] = '#';
            f--;
          }
      }
  }
}

void buildingadd(table_t *m,int N,int S,int E,int W,int ranX,int ranY){
  int i;
  srand(time(NULL));
  //Pokemon center will be on N S road and Pokemon mart will be on E W road
  //Check to see if building is being added to path.
  //also allow the marts to be added to other sides.
  int cloc = rand()%(ranX-2)+1;
  
  if(cloc == E|| cloc == W||m ->map[E+2][cloc] != ' '||m ->map[E+1][cloc+1] != ' ')
  {
    cloc += 4;
  }else if((cloc -1) == E|| (cloc - 1) == W)
  {
    cloc += 4;
  }


  int mloc = rand()%(ranY-2)+1;
  
  if(mloc == S || m ->map[mloc][N+2] != ' '||m ->map[mloc+1][N+1] != ' '){
    mloc -= 4; 
  }else if((mloc - 1) == S || (mloc - 1) == N )
  {
    mloc-= 4;
  }
  /*
  else if(mloc+1 == S){
    mloc -= 2;
  }else if(ranX == m->map[mloc][N+2])
  {
    mloc++;
  }
*/
  

  //printf("Center Location: %d\n Mart Location: %d\n",cloc,mloc);
//finds location for center
for(i = 0; i < 2; i++){
  m ->map[mloc+i][N+2] = 'C';
  m ->map[mloc+i][N+1] = 'C';
}
for(i = 0; i < 2; i++){
  m ->map[E+2][cloc+i] = 'M';
  m ->map[E+1][cloc+i] = 'M';
}
}

void grass_add(table_t *m){
  srand(time(NULL));
  int k;
  int tall_grass_x = rand()%3+28;
  int tall_grass_y = rand()%3+10;

  //printf("tall_grass_x %d, tall_grass_y %d\n",tall_grass_x,tall_grass_y);
  k = rand()% 2;
  //Adds tall grass region 1 which is top left corner
  while(tall_grass_y > 0 && tall_grass_x > 0)
  {
      if(m ->map[tall_grass_y][k] == ' ')
      {
        m ->map[tall_grass_y][k] = ':';
      }
      if(k == tall_grass_x){
        tall_grass_y--;
        tall_grass_x = rand()%21+40;
        k = rand() % 2;
      }
    k++;
  }
  //Adds tall grass region 2 which is bottom right corner
  tall_grass_y = 19;
  tall_grass_x = 79;
  //printf("second tall_grass_x %d, second tall_grass_y %d\n",tall_grass_x,tall_grass_y);
  k = 45;
  while(tall_grass_y >= 10 && tall_grass_x > 40)
  {
      if(m ->map[tall_grass_y][k] == ' ')
      {
        m ->map[tall_grass_y][k] = ':';
      }
      if(k == tall_grass_x){
        tall_grass_y--;
        tall_grass_x = 78;
        k = rand()%5+45;
      }
    k++;
  }
}

void clearing_builder(table_t *m){
  srand(time(NULL));
  int k;
  int x_clearing = rand()%3+28;
  int y_clearing = rand()%2+17;

  //printf("x_clearing %d, y_clearing %d\n",x_clearing,y_clearing);
  k = rand()% 2;
  //Adds clearing to bottom left corner
  while(y_clearing > 0 && x_clearing > 0)
  {
      if(m ->map[y_clearing][k] == ' ')
      {
        m ->map[y_clearing][k] = '.';
      }
      if(k == x_clearing){
        y_clearing--;
        x_clearing = rand()%21+40;
        k = rand() % 2;
      }
    k++;
  }
  y_clearing = rand()%3+10;
  x_clearing = 79;
  //printf("second x_clearing %d, second y_clearing %d\n",x_clearing,y_clearing);
  k = 45;
  while(y_clearing >= 0 && x_clearing > 40)
  {
      if(m ->map[y_clearing][k] == ' ')
      {
        m ->map[y_clearing][k] = '.';
      }
      if(k == x_clearing){
        y_clearing--;
        x_clearing = rand()%3+75;
        k = rand()%5+45;
      }
    k++;
  }

}

void trees_boulders(table_t *m){
  int i,j;
  srand(time(NULL));
  int ranNum = rand()%1;
  for(i = 0; i < 21; i++){
    for(j = 0; j < 80; j++){
      if(m ->map[i][j] == ' ')
      {
        if(ranNum == 0){
            m ->map[i][j] = '^';
        }else{
            m ->map[i][j] = '%';
        }
      }
      ranNum = rand()%2;
    }
  }
}

int main(int argc, char *argv[]){
  srand(time(NULL));
  //make sure that the roads have a 4x4 distance between them so that the buidlings have space between them.
  //prevent the roads from being close to the walls
  int N = rand()% 34+40;
  int S = rand()% 74 + 5;
  int E = rand()% 13 + 5;
  int W = rand()% 13 + 5;
  //Make sure the bounds of ranX and Y are properly defined so that it does not go of the map
  int ranX = rand()% 69 + 10;
  int ranY = rand()% 12 + 5;
  printf("RanY: %d E: %d",ranY,E);
  if(ranY-E < 3)
  {
    ranY+=4;
  }
  if(abs(N - ranX) <= 2){
    ranX+=2;
  }
  if(abs(E-W)|| abs(W-E) <= 2)
  {
    W++;
  }
   printf(" POST RanY: %d E: %d\n",ranY,E);
    //intitialize 2d array struct
    table_t* map = malloc(sizeof(table_t));
    int i,j;
    mapbuilder(map);
    pathbuilder(map,N,S,E,W,ranX,ranY);
    buildingadd(map,N,S,E,W,ranX,ranY);
    grass_add(map);
    clearing_builder(map);
    trees_boulders(map);
    for(i = 0; i < 21; i++){
        for(j = 0; j < 80; j++){
            printf("%c", map->map[i][j]);
        }
        printf("\n");
    }
}


TerrainGeneration.h
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
typedef struct table_t{
char map[21][80];
int N;
int S;
int E;
int W;
}table_t;
void mapbuilder(table_t *m){
  int i,j;
  for(i = 0; i < 21; i++){
      for(j = 0; j < 80; j++){
          if(i == 0|| i == 20 || j == 0|| j == 79)
          {
              m->map[i][j] = '%';

          }else{
              m->map[i][j] = ' ';
          }
      }
  }
}
int return_exits(table_t *m,char dir){
  switch(dir){
    case 'n':
      return m->N;
    case 's':
      return m->S;
    case 'e':
      return m->E;
    case 'w':
      return m->W;

  }
  //printf("N:%d S:%d E:%d W:%d\n",m->N,m->S,m->E,m->W);
}

void pathbuilder(table_t *m,int N,int S,int E,int W,int ranX,int ranY){

  int x_diff = E - W;
  int y_diff= N - S;
  //printf("X difference: %d, Y difference: %d\n",x_diff,y_diff);
  //printf("N:%d, S:%d\n",N,S);
  //printf("E:%d, W:%d, ranY: %d\n",E,W,ranY);
  m ->map[0][N] = '#';
  m ->map[20][S] = '#';
  m ->map[E][0] = '#';
  m ->map[W][79] = '#';
  int k = 79;
  int i,j;
  for(i = 0; i < 21; i++){
      for(j = 0; j < 80; j++){
          if(i == E && j <= ranX && j > 0){
              m->map[E][j] = '#';
          }
          if(E > W &&  x_diff>= 0){
              m ->map[E-x_diff][ranX] = '#';
              x_diff--;
          }else if(E < W && x_diff <= 0){
              m ->map[E-x_diff][ranX] = '#';
              x_diff++;
          }
          if(k > ranX)
          {
              m ->map[W][k] = '#';
              k--;
          }
      }
  }
  int f = 20;
  for(i = 0; i < 21; i++){
      for(j = 0; j < 80; j++){
          if(j == N && i <= ranY && i > 0){
               m ->map[i][N] = '#';
          }

          if(N > S && y_diff > 0){
               m ->map[ranY][N-y_diff] = '#';
                 y_diff--;

             }else if(N < S && y_diff < 0){
               m ->map[ranY][N-y_diff] = '#';
                 y_diff++;
             }

          if(f > ranY){
            m ->map[f][S] = '#';
            f--;
          }
      }
  }
}

void buildingadd(table_t *m,int N,int S,int E,int W,int ranX,int ranY,int rows,int cols){
  int i;
  srand(time(NULL));
  //Pokemon center will be on N S road and Pokemon mart will be on E W road
  //Check to see if building is being added to path.
  //also allow the marts to be added to other sides.
  //TODO ADD Probility for mart and center here
  double prob;
  if(rows > cols){
    prob = 100*(10/(double)abs(rows));
  }else{
    prob = 100*(10/(double)abs(cols));
  }
    int random = rand()%100+1;
    
  if(random > prob || (rows == 199 && cols == 199)){


    int cloc = rand()%(ranX-2)+1;

    if(cloc == E|| cloc == W||m ->map[E+2][cloc] != ' '||m ->map[E+1][cloc+1] != ' ')
    {
      cloc += 4;
    }else if((cloc -1) == E|| (cloc - 1) == W)
    {
      cloc += 4;
    }


    int mloc = rand()%(ranY-2)+1;

    if(mloc == S || m ->map[mloc][N+2] != ' '||m ->map[mloc+1][N+1] != ' '){
      mloc -= 4;
    }else if((mloc - 1) == S || (mloc - 1) == N )
    {
      mloc-= 4;
    }



    //printf("Center Location: %d\n Mart Location: %d\n",cloc,mloc);
  //finds location for center
  for(i = 0; i < 2; i++){
    m ->map[mloc+i][N+2] = 'C';
    m ->map[mloc+i][N+1] = 'C';
  }
  for(i = 0; i < 2; i++){
    m ->map[E+2][cloc+i] = 'M';
    m ->map[E+1][cloc+i] = 'M';
  }
  }


}

void grass_add(table_t *m){
  srand(time(NULL));
  int k;
  int tall_grass_x = rand()%3+28;
  int tall_grass_y = rand()%3+10;

  //printf("tall_grass_x %d, tall_grass_y %d\n",tall_grass_x,tall_grass_y);
  k = rand()% 2;
  //BUG REDUCE REGION SIZES
  //Adds tall grass region 1 which is top left corner
  while(tall_grass_y > 0 && tall_grass_x > 0)
  {
      if(m ->map[tall_grass_y][k] == ' ')
      {
        m ->map[tall_grass_y][k] = ':';
      }
      if(k == tall_grass_x){
        tall_grass_y--;
        tall_grass_x = rand()%21+40;
        k = rand() % 2;
      }
    k++;
  }
  //Adds tall grass region 2 which is bottom right corner
  tall_grass_y = 19;
  tall_grass_x = 79;
  //printf("second tall_grass_x %d, second tall_grass_y %d\n",tall_grass_x,tall_grass_y);
  k = 45;
  while(tall_grass_y >= 10 && tall_grass_x > 40)
  {
      if(m ->map[tall_grass_y][k] == ' ')
      {
        m ->map[tall_grass_y][k] = ':';
      }
      if(k == tall_grass_x){
        tall_grass_y--;
        tall_grass_x = 78;
        k = rand()%5+45;
      }
    k++;
  }
}

void clearing_builder(table_t *m){
  srand(time(NULL));
  int k;
  int x_clearing = rand()%3+28;
  int y_clearing = rand()%2+17;

  //printf("x_clearing %d, y_clearing %d\n",x_clearing,y_clearing);
  k = rand()% 2;
  //Adds clearing to bottom left corner
  while(y_clearing > 0 && x_clearing > 0)
  {
      if(m ->map[y_clearing][k] == ' ')
      {
        m ->map[y_clearing][k] = '.';
      }
      if(k == x_clearing){
        y_clearing--;
        x_clearing = rand()%21+40;
        k = rand() % 2;
      }
    k++;
  }
  y_clearing = rand()%3+10;
  x_clearing = 79;
  //printf("second x_clearing %d, second y_clearing %d\n",x_clearing,y_clearing);
  k = 45;
  while(y_clearing >= 0 && x_clearing > 40)
  {
      if(m ->map[y_clearing][k] == ' ')
      {
        m ->map[y_clearing][k] = '.';
      }
      if(k == x_clearing){
        y_clearing--;
        x_clearing = rand()%3+75;
        k = rand()%5+45;
      }
    k++;
  }

}

void trees_boulders(table_t *m){
  int i,j;
  srand(time(NULL));
  int ranNum = rand()%1;
  for(i = 0; i < 21; i++){
    for(j = 0; j < 80; j++){
      if(m ->map[i][j] == ' ')
      {
        if(ranNum == 0){
            m ->map[i][j] = '^';
        }else{
            m ->map[i][j] = '%';
        }
      }
      ranNum = rand()%2;
    }
  }
}

table_t* mapgen(table_t *x,int N,int S,int E,int W,int rows,int cols){
  srand(time(NULL));
  //make sure that the roads have a 4x4 distance between them so that the buidlings have space between them.
  //prevent the roads from being close to the walls
  //Make sure the bounds of ranX and Y are properly defined so that it does not go of the map
  int ranX = rand()% 69 + 10;
  int ranY = rand()% 12 + 5;
  //BUG fix ranY range so that the # will not be accross the bottom
 // printf("RanY: %d E: %d",ranY,E);
  if(ranY-E < 3)
  {
    ranY+=4;
  }
  if(abs(N - ranX) <= 2){
    ranX+=2;
  }
  /*
  if(abs(E-W)|| abs(W-E) <= 2)
  {
    W++;
  }
  */
  x->N = N;
  x->S = S;
  x->E = E;
  x->W = W;
  // printf(" POST RanY: %d E: %d\n",ranY,E);
    //intitialize 2d array struct
    int i,j;
    mapbuilder(x);
    pathbuilder(x,N,S,E,W,ranX,ranY);
    buildingadd(x,N,S,E,W,ranX,ranY,rows,cols);
    grass_add(x);
    clearing_builder(x);
    trees_boulders(x);
  return x;
}


Map2Map.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <string.h>
#include <ctype.h>
#include "TerrainGeneration.h"

typedef struct worldmap_t{
  table_t* world[399][399];
}worldmap_t;


int neighborCheck(worldmap_t *worldmap,int rows,int cols, char dir){


  srand(time(NULL));
  switch(dir){
    case 'n':
      if(worldmap->world[rows + 1][cols] != NULL){
        //takes in neighboring south as north
          return return_exits(worldmap->world[rows + 1][cols],'s');
      }else{
          return (rand()% 34+40);
      }
    case 's':
        if(worldmap->world[rows - 1][cols] != NULL){
      //takes neighboring north as south
          return return_exits(worldmap->world[rows - 1][cols],'n');
        }else{
          return (rand()% 74 + 5);
        }
    case 'e':
    if(worldmap->world[rows][cols + 1] != NULL){
      //takes neighboring west exit as east exit
          return return_exits(worldmap->world[rows][cols + 1],'w');
    }else{
      //randomize east exit
          return (rand()% 13 + 5);
    }
    case 'w':
    if(worldmap->world[rows][cols - 1] != NULL){
      //takes neighboring east as west exit
      return return_exits(worldmap->world[rows][cols - 1],'e');
    }else{
      //randomize west exit
      return (rand()% 13 + 5);
    }

  }



}
//this function will parse the integers out of input. first space is x second space is y
int parseX(char x[4]){
  printf("\n");
  int parsedX;
  sscanf(x, "%04d", &parsedX);
  return parsedX;
}
int parseY(char input[20],int second){
  int i;
  char posy[4];
   int count = 0;
  for(i = second; i < 20; i++)
  {
    if((isdigit(input[i]) || input[i] == '-')){
      posy[count] = input[i];

      count++;
    }
  }
  int parsed;
  sscanf(posy, "%04d", &parsed);
  return parsed;
}




void print_map(table_t* m, int x, int y){
  int i,j;
  for(i = 0; i < 21; i++){
        for(j = 0; j < 80; j++){
            printf("%c", m->map[i][j]);
        }
        printf("\n");
    }
    printf("(%d,%d)\n",x,y);
}
//checks if a neighbor exists and matches exits with current map. if it doesn't exist returns random exit


//make function called border which will take in cardinal direction map and will erase exit based on direction
table_t* border(table_t *b, char dir,int rows,int cols){
  int i,j;

  //BUG DOUBLE CHECK RANGES
  if(dir == 'n'){
    for(i = 0; i < 80; i++){
      b->map[0][i] = '%';
      if(cols == 0){
        b->map[i][80] = '%';
      }else if(cols = 399){
          b->map[i][0] = '%';
      }
    }
    }

  if(dir == 's'){
    for(i = 0; i < 80; i++){
      b->map[21][i] = '%';
      if(cols == 0){
        b->map[i][80] = '%';
      }else if(cols == 399){
        b->map[i][0] = '%';
      }
  }

  if(dir == 'e'){
    for(i = 0; i < 21; i++){
      b->map[i][0] = '%';
      if(rows == 0){
        b->map[0][i] = '%';
      }else if(rows == 399){
        b->map[21][i] = '%';
      }
    }
  }
  if(dir == 'w'){
    for(i = 0; i < 21; i++){
      b->map[i][80] = '%';
      if(rows == 0){
        b->map[0][i] = '%';
      }else if(rows == 399){
          b->map[21][i] = '%';
      }
    }
  }
}

  return b;
}

int main(int argc, char *argv[]){
  int i,j;
  srand(time(NULL));
  worldmap_t* worldmap = malloc(sizeof(worldmap_t));
  for( i = 0; i < 399; i++){
    for(j = 0; j < 399; j++){
      worldmap->world[i][j] = NULL;
    }
  }
  table_t* map = malloc(sizeof(table_t));

  int N = rand()% 34+40;
  int S = rand()% 74 + 5;
  int E = rand()% 13 + 5;
  int W = rand()% 13 + 5;
  //printf("E before mapgen %d\n",E);
  int xCoord = 0;
  int yCoord = 0;
  worldmap->world[199][199] = mapgen(map,N,S,E,W,199,199);
  print_map( worldmap->world[199][199],xCoord,yCoord);

  char exit = 0;
  int rows = 199;
  int cols = 199;
  int prevS = N;
  int prevN = S;
  int prevE = E;
  int prevW = W;
  //printf("returning exit %d\n",return_exits(worldmap->world[199][199],'w'));
  //printf("PREV W %d\n",prevW);





while(1)
{
    char c[20];
    fgets(c, 20, stdin);
    int count = 2;
    //use spaces as a buffer. when spaces == 1 count into x char string //Possibly use site bro gave me

    if(c[0] == 'q')
    {

        break;
    }


    if(c[0] == 'f'){


      char posx[4];
      char posy[4];

      int count = 0;
      int spaces = 0;
      int secondSpace = 0;
      for(i = 0; i < 20; i++)
      {
        if(c[i] == ' '){
          spaces++;
          if(spaces == 2){
            secondSpace = i;
          }
        }

        if(spaces == 1 && (c[i] == '-'||isdigit(c[i]))){
          posx[count] = c[i];
          count++;
        }

      }
    int X = parseX(posx);
    int Y = parseY(c,secondSpace);

    if(X > 199)
    {
      X = 199;
    }
    if(Y > 199){
      Y = 199;
    }
    if(X < 0){
      X = 0;
    }
    if(Y < 0){
      Y = 0;
    }
    rows = rows + Y;
    cols = cols + X;
    xCoord = X;
    yCoord = Y;
     //printf("%d , %d",parse(c,1),parse(c,2));
    if(worldmap->world[rows][cols] == NULL){
      N = neighborCheck(worldmap,rows,cols,'n');
      S = neighborCheck(worldmap,rows,cols,'s');
      E = neighborCheck(worldmap,rows,cols,'e');
      W = neighborCheck(worldmap,rows,cols,'w');
      table_t* map2 = malloc(sizeof(table_t));
      //check if it is teleported to a border on a side
      if(rows == 0 || cols == 0 || rows == 399 || cols == 399){
        worldmap->world[rows][cols] = border(mapgen(map2,N,prevS,E,W,rows,cols),'n',rows,cols);
      }else{
        worldmap->world[rows][cols] = mapgen(map2,N,prevS,E,W,rows,cols);
      }
      print_map(worldmap->world[rows][cols],xCoord,yCoord);
    }else{
      print_map( worldmap->world[rows][cols],xCoord,yCoord);
    }
    }





    //BUG switch south and north variable names because north is negative y coordinate direction but goes towards row == 399 while south is positive coording y direction but goes towards zero in 2d array
    if(c[0] == 's'){
      if(rows < 399){
        rows++;
        yCoord--;
      }
      //current issue is that for some reason, the value of north is being changed every iteration
      if(worldmap->world[rows][cols] == NULL && rows >= 0){
        //check neigbors and set current exit based on neighboring exit.

        //given south randomize east west and north
        N = neighborCheck(worldmap,rows,cols,'n');
        E = neighborCheck(worldmap,rows,cols,'e');
        W = neighborCheck(worldmap,rows,cols,'w');




        table_t* map2 = malloc(sizeof(table_t));
        if(rows == 0 || rows == 399){
          worldmap->world[rows][cols] = border(mapgen(map2,N,prevS,E,W,rows,cols),'s',rows,cols);
        }else{
          worldmap->world[rows][cols] = mapgen(map2,N,prevS,E,W,rows,cols);
        }
        prevS = N;
        prevN = prevS;
        prevE = E;
        prevW = W;
        print_map( worldmap->world[rows][cols],xCoord,yCoord);

      }else{

        print_map( worldmap->world[rows][cols],xCoord,yCoord);

      }

    }else if(c[0] == 'n'){
      //given north randomize west east and south.
      if(rows > 0){
        rows--;
        yCoord++;
      }
      if(worldmap->world[rows][cols] == NULL && rows != 399){

        S = neighborCheck(worldmap,rows,cols,'s');
        E = neighborCheck(worldmap,rows,cols,'e');
        W = neighborCheck(worldmap,rows,cols,'w');
        //later bug to be fixed

        table_t* map2 = malloc(sizeof(table_t));
        if(rows == 0 || rows == 399){
          worldmap->world[rows][cols] = border(mapgen(map2,prevN,S,E,W,rows,cols),'n',rows,cols);
        }else{
          worldmap->world[rows][cols] = mapgen(map2,prevN,S,E,W,rows,cols);
        }
        prevN = S;
        prevS = N;
        prevE = E;
        prevW = W;
        print_map( worldmap->world[rows][cols],xCoord,yCoord);
      }else{
        print_map( worldmap->world[rows][cols],xCoord,yCoord);
      }
    }else if(c[0] == 'e'){
      cols++;
      xCoord++;
      if(worldmap->world[rows][cols] == NULL){
        N = neighborCheck(worldmap,rows,cols,'n');
        S = neighborCheck(worldmap,rows,cols,'s');
        W = neighborCheck(worldmap,rows,cols,'w');

        //printf("PREV W %d\n",prevW);
        table_t* map2 = malloc(sizeof(table_t));
        if(cols == 399){
          worldmap->world[rows][cols] = border(mapgen(map2,N,S,prevW,W,rows,cols),'e',rows,cols);
        }else{
          worldmap->world[rows][cols] = mapgen(map2,N,S,prevW,W,rows,cols );
        }

        prevW = W;
        prevS = N;
        prevN = S;
        print_map( worldmap->world[rows][cols],xCoord,yCoord);
      }else{
        print_map( worldmap->world[rows][cols],xCoord,yCoord);
      }
    }else if(c[0] == 'w'){
      if(cols > 0){
        cols--;
        xCoord--;
      }
      if(worldmap->world[rows][cols] == NULL){
        N = neighborCheck(worldmap,rows,cols,'n');
        S = neighborCheck(worldmap,rows,cols,'s');
        E = neighborCheck(worldmap,rows,cols,'e');



        table_t* map2 = malloc(sizeof(table_t));
        if(cols == 0){
          worldmap->world[rows][cols] = border(mapgen(map2,N,S,E,prevE,rows,cols),'w',rows,cols);
        }else{
            worldmap->world[rows][cols] = mapgen(map2,N,S,E,prevE,rows,cols);
        }
        prevE = E;
        prevS = N;
        prevN = S;
        print_map( worldmap->world[rows][cols],xCoord,yCoord);
      }else{
        print_map( worldmap->world[rows][cols],xCoord,yCoord);
      }



    }
    //printf("Prev N: %d Prev S: %d Prev E: %d Prev W: %d\n",prevN,prevS,prevE,prevW);

    exit = c[0];
}

    free(map);
    free(worldmap);

}
